Bring etcd support for bucket DNS federation

harshavardhana opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Bring etcd support for bucket DNS federation
<!--- Describe your changes in detail -->

## Motivation and Context
- Supports centralized `config.json`
- Supports centralized `bucket` service records
  for client lookups

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
Manually, using multiple servers
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

harshavardhana committed:
Bring etcd support for bucket DNS federation

- Supports centralized `config.json`
- Supports centralized `bucket` service records
  for client lookups
- implement a new proxy forwarder

krisis commented on cmd/common-main.go:
```diff
@@ -40,6 +44,18 @@ func checkUpdate(mode string) {
+		if err := loadConfig(); err != nil {
+			if etcdc.IsKeyNotFound(err) {
+				fatalIf(newConfig(), "Unable to initialize minio config for the first time.")
+				log.Println("Created minio configuration file successfully at", globalEtcdClient.Endpoints())
```
process exits on `fatalIf`. `log.Println` after that doesn't have any effect.

harshavardhana replied:
> process exits on fatalIf. log.Println after that doesn't have any effect.

Only on error @krisis if not it prints the `log.Println`

krisis commented on cmd/bucket-handlers.go:
```diff
@@ -216,11 +217,29 @@ func (api objectAPIHandlers) ListBucketsHandler(w http.ResponseWriter, r *http.R
-		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
-		return
+
+	// If etcd, dns federation configured list buckets from etcd.
```
The comment could be more descriptive like,
`// if etcd is configured, list buckets from its key-value store. This would list only those buckets that were created after configuring etcd-coredns with minio instances.`

harshavardhana replied:
We need to fix this behavior @krisis so your comment might not be appropriate.

krisis commented on docs/federation/lookup/README.md:
```diff
@@ -0,0 +1,48 @@
+## Federation (Bucket Lookup)
+Bucket lookup federation requires two dependencies
+
+- etcd (for config, bucket SRV records)
```
we should expand on what information is being persisted in etcd.

harshavardhana replied:
what would be better to list?

krisis commented on docs/federation/lookup/README.md:
````diff
@@ -0,0 +1,48 @@
+minio server http://rack{5...8}.host{5...8}.domain.com/mnt/export{1...32}
+```
+
+In this configuration you can see `MINIO_ETCD_ENDPOINTS` points to the etcd backend which manages our
````
s/manages our/manages minio's/

krisis commented on docs/federation/lookup/README.md:
```diff
@@ -0,0 +1,48 @@
+# Federation
+There are primarily two types of federation
+
+- Bucket lookup
```
We should explain (at least briefly) what we mean when we say Bucket lookup and Bucket sharded.

krisis commented on cmd/server-main.go:
```diff
@@ -65,6 +65,13 @@ ENVIRONMENT VARIABLES:
 
+  BUCKET-DNS:
+     MINIO_DOMAIN:    To enable virtual-host-style requests.
+     MINIO_DOMAIN_IP: To enable virtual-host-style requests.
```
A note on sample values or allowed values would be useful.

harshavardhana replied:
Will document them separately in a doc. since this is just informative.

krisis reviewed (dismissed):
Some initial comments. I will be testing this feature locally and continue reviewing after that.

nitisht committed:
Add functionality to add old buckets to etcd on startup

Buckets already present on a Minio server before it joins a
bucket federated deployment will now be added to etcd during
startup. In case of a bucket name collision, admin is informed
via Minio server console message.

Added configuration migration for configuration stored in etcd
backend.

Also, environment variables are updated and ListBucket path style
request is no longer forwarded.

balamurugana commented on cmd/bucket-handlers.go:
```diff
@@ -106,6 +108,44 @@ func isBucketActionAllowed(action, bucket, prefix string, objectAPI ObjectLayer)
+// -- If IP of the entry doesn't match, this means entry is for another instance. Log an error to console.
+func initFederatorBackend() {
+	objLayer := newObjectLayerFn()
+	// List all buckets
```
remove this comment

harshavardhana replied:
// cc @nitisht

balamurugana commented on cmd/bucket-handlers.go:
```diff
@@ -106,6 +108,44 @@ func isBucketActionAllowed(action, bucket, prefix string, objectAPI ObjectLayer)
+				logger.LogIf(context.Background(), err)
+			}
+			return
+		}(bInfo.Name)
```
could you use meaningful names like `buckets`, `bucket`, `bucketName` etc?

harshavardhana replied:
// cc @nitisht

balamurugana commented on cmd/bucket-handlers.go:
```diff
@@ -106,6 +108,44 @@ func isBucketActionAllowed(action, bucket, prefix string, objectAPI ObjectLayer)
+		return
+	}
+	for _, bInfo := range b {
+		go func(bName string) {
```
Why no waitgroup used?

harshavardhana replied:
// cc @nitisht

balamurugana commented on cmd/common-main.go:
```diff
@@ -17,16 +17,21 @@
 	"strings"
 	"time"
 
+	etcdc "github.com/coreos/etcd/client"
```
better alias is `etcdclient`

balamurugana reviewed (commented):
partial review

harshavardhana committed:
Converge etcd functionality as part of quick.Config

nitisht committed:
Add etcd handling for web-handler methods

nitisht committed:
Add support for CopyObject across regions and multiple Minio IPs

This PR adds CopyObject support for objects residing in buckets
in different Minio instances (where Minio instances are part of
a federated setup).

Also, added support for multiple Minio domain IPs. This is required
for distributed deployments, where one deployment may have multiple
nodes, each with a different public IP.

nitisht committed:
Use random host from among multiple hosts to create requests

Also use hosts passed to Minio startup command to populate IP
addresses if MINIO_PUBLIC_IPS is not set.

vadmeste commented on cmd/api-errors.go:
```diff
@@ -864,34 +866,42 @@ func toAPIErrorCode(err error) (apiErr APIErrorCode) {
+		apiErr = ErrOperationTimedOut
+	}
+
+	if apiErr != ErrNone {
```
I don't think this is correct because no code sets apiErr to ErrNone already. You probably need to initialize apiErr to ErrNone in the beginning of the function.

harshavardhana replied:
ApiErr default type value is None so it will work.

vadmeste commented on cmd/generic-handlers.go:
```diff
@@ -611,6 +618,75 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+// To forward the path style requests on a bucket to the right
+// configured server, bucket to IP configuration is obtained
+// from centralized etcd configuration service.
+type bucketForwardingHandler struct {
```
@harshavardhana, we probably can return a temporary redirect here https://docs.aws.amazon.com/AmazonS3/latest/dev/Redirects.html#TemporaryRedirection in case of someone puts a firewall between federators  but we can use this at the moment

harshavardhana replied:
This may not be needed any specific reason why we would send this error?

vadmeste reviewed (dismissed):
LGTM & tested

krisis commented on cmd/common-main.go:
```diff
@@ -123,11 +151,46 @@ func handleCommonEnvVars() {
+		var err error
+		globalEtcdClient, err = etcd.New(etcd.Config{
+			Endpoints: etcdEndpoints,
+			Transport: &http.Transport{
```
we could use the existing `http.Transport`

harshavardhana replied:
done

krisis commented on cmd/common-main.go:
```diff
@@ -123,11 +151,46 @@ func handleCommonEnvVars() {
+		logger.FatalIf(err, "Unable to initialize etcd with %s", etcdEndpoints)
+	}
+
+	globalDomainName, ok = os.LookupEnv("MINIO_DOMAIN")
```
suggestion: This can be simplified.
`globalDomainName, globalIsEnvDomainName = os.LookupEnv("MINIO_DOMAIN")`

harshavardhana replied:
done

krisis commented on cmd/config-versions.go:
```diff
@@ -585,6 +586,8 @@ type serverConfigV22 struct {
 // IMPORTANT NOTE: When updating this struct make sure that
 // serverConfig.ConfigDiff() is updated as necessary.
 type serverConfigV23 struct {
+	quick.Config `json:"-"` // ignore interfaces
```
this would move to `serverConfigV24` on rebase

harshavardhana replied:
done

krisis commented on cmd/endpoint.go:
```diff
@@ -493,3 +496,22 @@ func GetRemotePeers(endpoints EndpointList) []string {
+	_, dok := os.LookupEnv("MINIO_DOMAIN")
+	_, eok := os.LookupEnv("MINIO_ETCD_ENDPOINTS")
+	_, iok := os.LookupEnv("MINIO_PUBLIC_IPS")
+	if dok && eok && !iok {
```
why do we depend on `dok` and `eok` to update domain IPs from endpoint list?

harshavardhana replied:
globalDomainIPs is only needed for federation that's why.

krisis commented on cmd/generic-handlers.go:
```diff
@@ -611,6 +617,73 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+func setBucketForwardingHandler(h http.Handler) http.Handler {
+	fwd := handlers.NewForwarder(&handlers.Forwarder{
+		PassHost:     true,
+		RoundTripper: NewCustomHTTPTransport(),
```
do we need a new http transport? Could we reuse an existing transport?

harshavardhana replied:
it is already using existing transport, it just initializes a new one while registering the handler and reuses in all future operations.

krisis commented on cmd/object-handlers.go:
```diff
@@ -536,13 +538,73 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
+					return
+				}
+				// Close writer explicitly to indicate data has been written
+				defer srcInfo.Writer.Close()
```
we don't need the defer here.

harshavardhana replied:
done

vadmeste reviewed (approved):
LGTM & tested
